tests/test_handshake.py

```diff
@@ -107,5 +108,50 @@ def test_client_handshake_data(self):
                 assert future()
                 return
 
+    @unittest.skipUnless(version_tuple >= (3, 11, -1), "requires PyMongo 3.11")
+    def test_client_handshake_saslSupportedMechs(self):
+        server = MockupDB()
+        server.run()
+        self.addCleanup(server.stop)
+
+        primary_response = OpReply('ismaster', True,
+                                   minWireVersion=2, maxWireVersion=6)
+        client = MongoClient(server.uri,
+                             username='username',
+                             password='password')
+
+        self.addCleanup(client.close)
+
+        # New monitoring sockets send data during handshake.
+        heartbeat = server.receives('ismaster')
+        heartbeat.ok(primary_response)
+
+        future = go(client.db.command, 'whatever')
+        for request in server:
+            if request.matches('ismaster'):
+                if request.client_port == heartbeat.client_port:
+                    # This is the monitor again, keep going.
+                    request.ok(primary_response)
+                else:
+                    # Handshaking a new application socket should send
+                    # saslSupportedMechs and speculativeAuthenticate.
+                    self.assertEqual(request['saslSupportedMechs'],
+                                     'admin.username')
+                    self.assertIn(
+                        'saslStart', request['speculativeAuthenticate'])
+                    auth = {'conversationId': 1, 'done': False,
+                            'payload': b'r=wPleNM8S5p8gMaffMDF7Py4ru9bnmmoqb0'
+                                       b'1WNPsil6o=pAvr6B1garhlwc6MKNQ93ZfFky'
+                                       b'tXdF9r,s=4dcxugMJq2P4hQaDbGXZR8uR3ei'
+                                       b'PHrSmh4uhkg==,i=15000'}
+                    request.ok('ismaster', True,
+                               saslSupportedMechs=['SCRAM-SHA-256'],
+                               speculativeAuthenticate=auth,
+                               minWireVersion=2, maxWireVersion=6)
+                    with self.assertRaises(OperationFailure):
```

Comment:
The command fails because we don't actually support the server side of scram auth and the speculativeAuthenticate response doesn't match the client's:
```
OperationFailure: Server returned an invalid nonce.
```

Reply:
So we can only check that we initiate the SASL conversation correctly and that's it?

Reply:
Yep. Unless we implement the full server-side auth conversation (which would be overkill).